llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp

```diff
@@ -44,6 +44,18 @@ using ExtInstList =
 
 namespace {
 
+uint64_t getUnsignedConstantValueFromReg(llvm::Register Reg,
+                                         const llvm::MachineRegisterInfo &MRI) {
+  llvm::SPIRVType *ConstTy = MRI.getVRegDef(Reg);
+  assert(ConstTy && ConstTy->getOpcode() == llvm::SPIRV::ASSIGN_TYPE &&
+         ConstTy->getOperand(1).isReg());
```

Comment:
A bit weird, but I suppose this always run after we emitted the assign-type instructions, meaning we have this additional walk to do :/
Should the function do something like:
```
if (getOpcode() == ASSIGN_TYPE)
     return getUnsignedConstantValueFromReg(getOperand(1).getReg())
assert(getOpcode() == G_CONSTANT)
return getOperand(1).getCImm()
```

Reply:
This is even more complicated than that. Please have a look at https://github.com/llvm/llvm-project/blob/8535758efac9d7bb90fd65f883f46dbdaa8e9aa5/llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp#L1842-L1887

But if we are going to get a value from integer constant only it would be indeed something like

```
if (SPIRVType *OpDef = MRI->getVRegDef(OpReg)) {
  if (OpDef->getOpcode() == SPIRV::ASSIGN_TYPE &&
      OpDef->getOperand(1).isReg()) {
    if (SPIRVType *RefDef = MRI->getVRegDef(OpDef->getOperand(1).getReg()))
      OpDef = RefDef;
  }
  if (OpDef->getOpcode() == TargetOpcode::G_CONSTANT)
    return OpDef->getOperand(1).getCImm();
}
report_fatal_error(...);
```

Reply:
I made the helper function more robust by removing the assert for assign type as you suggested. However, I did not put that change outside if the helper function because I want to reduce the amount of repeated code.

On that note, I remembered that I copied the code in that function from another spot in the same file. So I replaced the original code with a call to the new function as well.

Reply:
Actually, I went looking more into the file and found an existing function that does what I want `foldImm`. I replaced all uses of my new function with a call to that function.